framework/src/mesh/ConcentricCircleMesh.C

```diff
@@ -0,0 +1,660 @@
+      node_id++;
+    }
+  }
+
+  // for adding the outer nodes of the square
+  Real current_radius = 0.0;
+
+  for (unsigned layers = 0; layers < total_concentric_circles.size(); ++layers)
+  {
+    current_radius = total_concentric_circles.at(layers);
+    for (unsigned num_outer_nodes = 0; num_outer_nodes <= _num_sectors; ++num_outer_nodes)
+    {
+      const Real x = current_radius * std::cos(num_outer_nodes * d_angle);
+      const Real y = current_radius * std::sin(num_outer_nodes * d_angle);
+      nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+      node_id++;
+    }
+  }
+
+  // adding nodes for the unit cell of fuel assembly.
+  if (_pitch != 0.0)
+  {
+    Real current_radius_moderator = 0.0;
+    for (unsigned i = 1; i <= _rings.at(_rings.size() - 1); ++i)
+    {
+      current_radius_moderator =
+          _radii.at(_radii.size() - 1) +
+          i * (_pitch / 2 - _radii.at(_radii.size() - 1)) / (_rings.at(_rings.size() - 1) + 1);
+      total_concentric_circles.push_back(current_radius_moderator);
+      for (unsigned num_outer_nodes = 0; num_outer_nodes <= _num_sectors; ++num_outer_nodes)
+      {
+        const Real x = current_radius_moderator * std::cos(num_outer_nodes * d_angle);
+        const Real y = current_radius_moderator * std::sin(num_outer_nodes * d_angle);
+        nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+        node_id++;
+      }
+    }
+
+    for (unsigned j = 0; j < _num_sectors / 2 + 1; ++j)
+    {
+      const Real x = _pitch / 2;
+      const Real y = _pitch / 2 * std::tan(j * d_angle);
+      nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+      node_id++;
+    }
+
+    for (unsigned i = 0; i < _num_sectors / 2; ++i)
+    {
+      const Real x = _pitch / 2 * std::cos((i + _num_sectors / 2 + 1) * d_angle) /
+                     std::sin((i + _num_sectors / 2 + 1) * d_angle);
+      const Real y = _pitch / 2;
+      nodes[node_id] = mesh.add_point(Point(x, y, 0.0), node_id);
+      node_id++;
+    }
+  }
+
+  // adding elements
+  int index = 0;
+  int limit = 0;
+  int standard = static_cast<int>(_num_sectors);
+
+  // This is to set the limit for the index
+  if (standard > 4)
+  {
+    int additional_term = 0;
+    int counter = standard;
+    while (counter > 4)
+    {
+      counter = counter - 2;
+      additional_term = additional_term + counter;
+    }
+    limit = standard + additional_term;
+  }
+  else if (standard == 4)
+  {
+    limit = standard;
+  }
+
+  // SubdomainIDs set up
+  std::vector<unsigned int> subdomainIDs;
+  for (unsigned int i = 0; i < _rings.size(); ++i)
+  {
+    for (unsigned int j = 0; j < _rings.at(i); ++j)
+    {
+      subdomainIDs.push_back(i + 1);
+    }
+  }
+  if (_pitch != 0.0)
+    subdomainIDs.push_back(subdomainIDs.at(subdomainIDs.size() - 1));
+  // adding elements in the square
+  while (index <= limit)
+  {
+    Elem * elem = mesh.add_elem(new Quad4);
```

Comment:
good start here with Quad4, but we should also think how to accommodate second order meshes. Any thoughts on that?